### ISSUE TITLE
Mixed Content problem

### ISSUE THREAD

**Issue by marentwickler** [nymtech/nym#400]
hello, webassembly Mixed Content problem, by loading Nym over HTTPS(based on the SSL cert), the connection has been blocked, while the Nym-Gateway endpoint(https://qa-directory.nymtech.net) must be available over WSS. but the actual WebSocket endpoint is available only over WS. is there a way to fix this Problem? 
Thanks

**Comment by futurechimp** [comment]
Hi @marentwickler can you provide a little more info about what you're trying to do? 

The websocket should be serving OK via WSS as far as I'm aware - at least I don't get a mixed-content warning in Firefox. The code that's loading it is:

```
conn = new WebSocket("wss://" + document.location.host + "/ws");
```

I could easily be wrong, but if I remember correctly I had set up websockets to serve via an NGINX proxy on that box which should provide TLS on the connection properly. 

Are you saying you get a warning when you hit https://qa-directory.nymtech.net and the JavaScript there tries to load the websocket in the browser? Or are you building your own client application?

**Comment by marentwickler** [comment]
First of all, thank you for your answer, I use  JavaScript and I tried to load the websocket in the browser, and  get a mixed-content warning in google-chrome:
![Screen Shot 2020-10-27 at 12 25 07](https://user-images.githubusercontent.com/11528983/97297028-c2875a00-1851-11eb-8cfa-19359e02da34.png)
and I have tried to change the endpoint in mods.rs by wasm-utils/ for testing:
let url='wss://178.79.164.25:9000/'
let ws = WebSocket::new(url)
 but without success.

**Comment by futurechimp** [comment]
Ok, so if I understand the setup, you are attempting to write a server-based nodejs application, and you want it to be able to send messages through the mixnet.

Are you running the Nym client on the same machine that the nodejs app is running on? 

You'll will want to initialize a Nym native clienr and send messages to the websocket on `localhost`, so there won't be any certificates or HTTPS needed, and you won't get any mixed content warnings. Instructions for the basics of how to do that are here:

https://nymtech.net/docs/build-peapps/native-client/

So basically you connect to the localhost websocket and send defined messages up it. There is a JavaScript example at:

https://github.com/nymtech/nym/tree/develop/clients/native/examples/js-examples/websocket

Is that sort of what you're trying to do?

**Comment by marentwickler** [comment]
Hello, I try to run webassembly/js-example  in https site like: https://xxxxxx:port, not native example , but i get the error from Mixed-content in google chrome. I use the same directory "https://qa-directory.nymtech.net", I will use this directory without setup Nym locally

**Comment by futurechimp** [comment]
Ah, now I understand.

Ok, I've tried to set up a project myself to see if I can reproduce whatever behaviour you're seeing and fix it, but I don't normally use Node and I'm having trouble getting the basic module requires working. 

Do you happen to have a repo that I can download this from and try it out? That way I can probably figure out what the problem is and get back to you with a solution to it tomorrow. 

Or if you can tell me why I get "MODULE_NOT_FOUND" from:

```
// package.json

{
  "name": "nym-nodejs-example",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "start": "node ./src/index.js"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@nymproject/nym-client-wasm": "^0.8.1"
  }
}
```

```javascript
// src/index.js

const { NymClient } = require('@nymproject/nym-client-wasm');
console.log("HI");
```

I have never tried to use an `@` in a node module before, is that syntax wrong? It seems like it should work.

**Comment by marentwickler** [comment]
Hi, I use actual repository https://github.com/nymtech/nym/tree/develop/clients/webassembly/js-example
 i use the package
```
// package.json
        {
           "scripts": {
            "start": "webpack  && node  server.js"
            },
            "dependencies": {
                "express": "^4.17.1",
                "@nymproject/nym-client-wasm": "file:../pkg"
            }
        }
```
,
and I have created Nodejs server with SSL certificate, the javascript code in client side shows in this form:
```
        Main = async () => {
                                try {
                                        const { NymClient,set_panic_hook } = await  import('@nymproject/nym-client-wasm');
                                       ......
                                        const directory = "https://qa-directory.nymtech.net"; 
                                    //"http://localhost:8080";
                                        client = new NymClient(directory);
                                        .......
                                        .......
                                        .....
                                } catch(err) {
                                        console.log(`Unexpected error in loadWasm. [Message: ${err.message}]`);
                                    }
                             } 
                  ...........
```
and it will be called in html page with <script  src="./index.js"></script>
the script in  site with HTTP: http://xxxxxx:port,  works fine, and i can send and receive the messages over Nym- gateway perfect, but with HTTPS:  https://xxxxxx:port it  is Mixed Content problematically in google chrome.

**Comment by marentwickler** [comment]
why you get "MODULE_NOT_FOUND"? while you will use "@nymproject/nym-client-wasm" in server side, what  is difficult with files structure in  pkg/package.json, you can use it only in client side.

**Comment by futurechimp** [comment]
> why you get "MODULE_NOT_FOUND"? while you will use "@nymproject/nym-client-wasm" in server side, what is difficult with files structure in pkg/package.json, you can use it only in client side.

Ok, I'll stop fighting with that then :). 

Back to the main problem, let's see if we can figure it out.

The client needs to have a valid Nym directory server URL when you call `client = new NymClient(directory);` - the client will then retrieve information from that directory about the state of the network, and use the retrieved information to figure out what mixnodes exist. The client will then automatically pick routes through the mix network, build encrypted Sphinx packets to send through the network, and then send the packets to the gateway. Note: the gateway is different than the directory.

So for instance, if you use `https://qa-directory.nymtech.net` as your server (that's our testnet), then the client will retrieve information from https://qa-directory.nymtech.net/api/presence/topology and use it to construct routes. The `mixnodes` array in there has mixnodes that anonymously relay traffic. The `gateways` array contains a list of Nym gateways, which act as a simple storage and entry point for the mixnodes. The gateways allow clients to be offline and then come online and retrieve packets that arrived when they were disconnected. 

It seems like you're trying to make an Express web application that you can make requests to, initializing a client from within the Express application, and then hitting the Express application in Chrome. 

I'll try and play back how I understand what you're doing, please correct me if I get something wrong. 

What I think *should* happen is that the browser sends whatever it wants to Express, and that part is entirely between Chrome and Express. 

Then Express code has a Nym client inside it, which it uses to send information from Express to the Nym gateway (which is different than the Nym directory).

None of that happens in the browser (?) so I wouldn't expect that Chrome would show a mixed-content warning here. In theory your code would make a headless request from Express to https://qa-directory.nymtech.net, get back network topology information, format a Sphinx packet, and send that Sphinx packet to the Nym gateway (choosing one of the gateways from the list returned in the `gateways` array).

I think we will be able to figure out quite quickly what's going on if I can see the whole application. I have tried using your `server.js` and `package.json` to hack around with the `js-example` but I'm still not totally clear on how your code is structured so I can't reproduce what you're seeing. 

Can you put up a Git repo with your code inside? Just copying and pasting the `js-example` into a new repo, or submitting a pull request with what you've got locally, would do the trick.

**Comment by tommyv1987** [comment]
I'm going to close this issue as the reporter has responded back to the previous message, this issue can be bought back to life in and when the reporter responds.
